### PR TITLE
Trivial Serialization Tweaks

### PR DESCRIPTION
**Pull request by TheBlueMatt**
These are the serialization prereqs for #1897.

### CODE REVIEW

File: lightning/src/util/ser.rs [hunk 1]
```diff
@@ -381,6 +383,35 @@ impl Readable for BigSize {
...
+/// length integer here which is backwards-compatible but treats 0xffff as "read eight more bytes".
+struct U16Or64(pub u64);
+impl Writeable for U16Or64 {
+	#[inline]
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
+		if self.0 < 0xffff {
+			(self.0 as u16).write(writer)
+		} else {
+			0xffffu16.write(writer)?;
+			(self.0 - 0xfffe).write(writer)
```
Comment (arik-so) [inline lightning/src/util/ser.rs:398]: I think you may wanna comment on the algorithm here, explaining why we then subtract one byte less than the value we just wrote. Same thing for the deserialization logic below.

**Review by valentinewallace — APPROVED** [review]
ACK after Arik's comment

File: lightning/src/util/ser.rs [hunk 2]
```diff
@@ -383,6 +383,35 @@ impl Readable for BigSize {
 	}
 }
 
+/// The lightning protocol uses u16s for lengths in most cases. As our serialization framework
+/// primarily targets that, we must as well. However, because we may serialize objects that have
+/// more than 65K entries, we need to be able to store larger values. Thus, we define a variable
+/// length integer here which is backwards-compatible but treats 0xffff as "read eight more bytes".
+struct U16Or64(pub u64);
+impl Writeable for U16Or64 {
```
Comment (jkczyz) [inline lightning/src/util/ser.rs:391]: This kinda looks like `U160r64`. Would prefer to name this either `CollectionLength` or at least `U16OrU64` which is a little easier to read.

File: lightning/src/util/ser.rs [hunk 3]
```diff
@@ -694,18 +723,23 @@ macro_rules! impl_for_vec {
 impl Writeable for Vec<u8> {
 	#[inline]
 	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
-		(self.len() as u16).write(w)?;
+		U16Or64(self.len() as u64).write(w)?;
```
Comment (jkczyz) [inline lightning/src/util/ser.rs:726]: Should be concerned with writing an incompatible length for an LN message? Or are we leaving that to the caller to enforce / assume will never happen?

File: lightning/src/util/ser.rs [hunk 4]
```diff
@@ -381,6 +383,38 @@ impl Readable for BigSize {
 	}
 }
 
+/// The lightning protocol uses u16s for lengths in most cases. As our serialization framework
+/// primarily targets that, we must as well. However, because we may serialize objects that have
+/// more than 65K entries, we need to be able to store larger values. Thus, we define a variable
+/// length integer here which is backwards-compatible but treats 0xffff as "read eight more bytes".
+///
+/// To ensure we only have one valid encoding per value, we add 0xffff to values written as eight
+/// bytes, representing 0xffff as zero-eight-bytes, which 0xfffe is 0xfffe-two-bytes.
```
Comment (valentinewallace) [inline lightning/src/util/ser.rs:392]: Why is 0xfffe being brought into this now...? This sentence confuses me, I think the grammar is off

File: lightning/src/util/ser.rs (hunk 2 above)
Comment (arik-so) [reply lightning/src/util/ser.rs]: Strongly agree with Jeff. I only ever read it as 160r64 and didn't realize there was an `Or` in there until Jeff's comment. I wasn't sure what either `160` or `r` meant, but didn't even think to ask. The `O` of `Or` remains rather unfortunate.

File: lightning/src/util/ser.rs [hunk 5]
```diff
@@ -381,6 +383,39 @@ impl Readable for BigSize {
...
 }
 
+/// The lightning protocol uses u16s for lengths in most cases. As our serialization framework
+/// primarily targets that, we must as well. However, because we may serialize objects that have
+/// more than 65K entries, we need to be able to store larger values. Thus, we define a variable
+/// length integer here which is backwards-compatible but treats 0xffff as "read eight more bytes".
+///
+/// To ensure we only have one valid encoding per value, we add 0xffff to values written as eight
+/// bytes. Thus, 0xfffe is serialized as 0xfffe, whereas 0xffff is serialized as
+/// 0xffff0000000000000000 (i.e. read-eight-bytes then zero).
```
Comment (arik-so) [inline lightning/src/util/ser.rs:393]: this mixed endianness is so confusing (0xfffe being big endian, but 0xffff0000000000000000 being little endian)

Comment (arik-so) [reply lightning/src/util/ser.rs:393]: How can the second number be big endian?

Comment (arik-so) [reply lightning/src/util/ser.rs:393]: oh, I didn't realize that the 0xffff portion is completely discarded, sorry.

Comment (arik-so) [reply lightning/src/util/ser.rs:393]: so we don't wanna do anything clever where we could encode up to `u64::max-0xffff`, and keep a total max encoding length of u64?

Comment (arik-so) [reply lightning/src/util/ser.rs:393]: yeah, probably not worth it, disregard this entire comment chain.

**Comment by valentinewallace** [comment]
LGTM after squash

File: lightning/src/util/ser.rs (hunk 2 above)
Comment (jkczyz) [reply lightning/src/util/ser.rs]: lol I had to Google it before realizing it was not a zero. Not sure it's worth putting the implementation details into the name, tbh.

File: lightning/src/util/ser.rs [hunk 6]
```diff
@@ -381,6 +383,39 @@ impl Readable for BigSize {
 	}
 }
 
+/// The lightning protocol uses u16s for lengths in most cases. As our serialization framework
+/// primarily targets that, we must as well. However, because we may serialize objects that have
+/// more than 65K entries, we need to be able to store larger values. Thus, we define a variable
+/// length integer here which is backwards-compatible but treats 0xffff as "read eight more bytes".
```
Comment (jkczyz) [inline lightning/src/util/ser.rs:389]: s/which is backwards-compatible/that is backwards-compatible for < 0xffff

**Review by arik-so — APPROVED** [review]
I blame Github's and my diff tool's fonts.

File: lightning/src/util/ser.rs (hunk 6 above)
Comment (jkczyz) [reply lightning/src/util/ser.rs]: Note the last part that says it's backwards compatible only for < 0xffff.

Comment (arik-so) [reply lightning/src/util/ser.rs]: I think if somebody is at exactly 0xffff, they're about to run into problems anyway.

**Review by jkczyz — APPROVED** [review]

File: lightning/src/util/ser.rs [hunk 7]
```diff
@@ -381,6 +383,40 @@ impl Readable for BigSize {
...
+	}
+}
+
+impl Readable for CollectionLength {
+	#[inline]
+	fn read<R: Read>(r: &mut R) -> Result<Self, DecodeError> {
+		let mut val: u64 = <u16 as Readable>::read(r)? as u64;
+		if val == 0xffff {
+			val = <u64 as Readable>::read(r)?
+				.checked_add(0xffff).ok_or(DecodeError::InvalidValue)?;
```
Comment (jkczyz) [inline lightning/src/util/ser.rs:414]: Oh, wait, can we forgo the addition here and the subtraction in the write by simply ignoring the first byte on read (when `val == 0xffff`) and writing the full `u64` after `0xffff` in the write (when `val >= 0xffff`)?

**Review by valentinewallace — APPROVED** [review]

**Pull request merged**